sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/datastore/DataStoreV1Table.java

```diff
@@ -0,0 +1,483 @@
+          case GEO_POINT_VALUE:
+          default:
+            throw new IllegalStateException(
+                "No conversion exists from type: "
+                    + val.getValueTypeCase().name()
+                    + " to Beam type. Supported types are: "
+                    + SUPPORTED_VALUE_TYPES.toString());
+        }
+      }
+
+      /**
+       * Converts all properties of an {@code Entity} to Beam {@code Row}.
+       *
+       * @param schema Target row {@code Schema}.
+       * @param values A map of property names and values.
+       * @return resulting Beam {@code Row}.
+       */
+      private Row extractRowFromProperties(Schema schema, Map<String, Value> values) {
+        Row.Builder builder = Row.withSchema(schema);
+        // It is not a guarantee that the values will be in the same order as the schema.
+        // Maybe metadata:
+        // https://cloud.google.com/appengine/docs/standard/python/datastore/metadataqueries
+        // TODO: figure out in what order the elements are in (without relying on Beam schema).
+        for (Schema.Field field : schema.getFields()) {
+          Value val = values.get(field.getName());
+          builder.addValue(convertValueToObject(field.getType(), val));
+        }
+        return builder.build();
+      }
+    }
+  }
+
+  /**
+   * A {@code PTransform} to perform a conversion of {@code PCollection<Row>} to {@code
+   * PCollection<Entity>}.
+   */
+  public static class RowToEntity extends PTransform<PCollection<Row>, PCollection<Entity>> {
+    private final Supplier<String> keySupplier;
+    private final String kind;
+    private final String keyField;
+
+    private RowToEntity(Supplier<String> keySupplier, String kind, String keyField) {
+      this.keySupplier = keySupplier;
+      this.kind = kind;
+      this.keyField = keyField;
+    }
+
+    @Override
+    public PCollection<Entity> expand(PCollection<Row> input) {
+      if (input.getSchema().getFieldNames().contains(keyField)
+          && !input.getSchema().getField(keyField).getType().getTypeName().equals(TypeName.BYTES)) {
+        throw new IllegalStateException(
+            "Field `"
+                + keyField
+                + "` should of type `VARBINARY`. Please change the type or specify a field to write the KEY value from via TableProperties.");
+      }
+      return input.apply(ParDo.of(new RowToEntityConverter()));
+    }
+
+    /**
+     * Create a PTransform instance.
+     *
+     * @param keyField Row field containing a serialized {@code Key}, must be set when using user
+     *     specified keys.
+     * @param kind DataStore `Kind` data will be written to (required when generating random {@code
+     *     Key}s).
+     * @return {@code PTransform} instance for Row to Entity conversion.
+     */
+    public static RowToEntity create(String keyField, String kind) {
+      LOGGER.info(
+          "VARBINARY field with the KEY was not specified, using default value: `"
```

Comment:
I'm confused. This error message is printed without even checking `keyField`.

Reply:
Simplified and moved this `info` message.
Should be displayed when `keyField` is one of the schema fields.